Fixing the threading issue in the syslog monitor

alesnovak-s1 opened the pull request:
Requests are read in a thread pool first to release the socket fd as soon as possible to ensure higher throughput.
The data is handled in a separate thread pool.

https://sentinelone.atlassian.net/browse/DTIN-3231

tomaz-s1 commented on scalyr_agent/builtin_monitors/thread_pool.py:
```diff
@@ -0,0 +1,59 @@
+import threading
```
Let's please add ASF license header to this file.

tomaz-s1 commented on scalyr_agent/builtin_monitors/syslog_monitor.py:
```diff
@@ -18,6 +18,10 @@
 from __future__ import absolute_import
```
It would be good to add a docstring to this module to explain to the developers how request processing works and that we use a thread pool for processing requests and reading data from the socket.

tomaz-s1 commented on scalyr_agent/builtin_monitors/syslog_monitor.py:
```diff
@@ -570,77 +576,113 @@ class SyslogUDPHandler(six.moves.socketserver.BaseRequestHandler):
         }
-        self.server.syslog_handler.handle(data, extra)
 
+        self.__request_processing_executor.submit(
```
Since `handle` is now called inside a thread, do we need any kind of locking around this method?

It appears we already have a lock in https://github.com/scalyr/scalyr-agent-2/blob/888447321333b7f817739cb06806a461188b3ba7/scalyr_agent/builtin_monitors/syslog_monitor.py#L1583 for file creation related reasons and all `handle()` method does is pass data to the logger (https://github.com/scalyr/scalyr-agent-2/blob/888447321333b7f817739cb06806a461188b3ba7/scalyr_agent/builtin_monitors/syslog_monitor.py#L1738).

I believe we already have a lock in our custom logger methods, but it would be good to confirm there are no race conditions - starting a monitor and sending a lot of data to in in parallel manner and verifying that syslog monitor log file is correct aka that data is not interleaved.

tomaz-s1 replied:
Thanks for the clarification :)

tomaz-s1 commented on scalyr_agent/builtin_monitors/thread_pool.py:
```diff
@@ -0,0 +1,59 @@
+class ExecutorMixIn:
+    def __init__(self):
+        # Using 4 threads for processing requests, because of GIL and CPU bound tasks higher number would not help process the data faster.
+        self._request_processing_executor = ThreadPoolExecutorFactory.get_singleton("request_processing_executor", max_workers=4)
```
Do we want to make `max_workers` configurable?

Perhaps it's a good idea in case someone hits a race condition or similar so we can provide them with a workaround to set this value to `1`.

alesnovak-s1 replied:
This was a laziness on my part, I will make it configurable :)

tomaz-s1 commented on scalyr_agent/monitors_manager.py:
```diff
@@ -208,6 +210,10 @@ def stop_manager(self, wait_on_join=True, join_timeout=5):
         start_time = time.time()
 
+        log.info("Stopping shared thread pools")
+        ThreadPoolExecutorFactory.shutdown(wait=False, cancel_futures=True)
```
To me it seems like that this should live on the syslog monitor `stop()` method and not inside the MonitorsManager itself.

alesnovak-s1 replied:
The thread pools are shared across the syslog monitors. That's why it needs to be here.

tomaz-s1 replied:
Ah, ok, makes sense, thanks for clarifying.

Could this cause any issues since each monitor is running in a separate thread already?

alesnovak-s1 replied:
That's a good question.
ThreadPool.submit (this method is thread safe) adds a callable item with parameters to a queue and starts a new thread and adds it to an internal pool if num_threads < self._max_workers.
What I was worried about is creating and starting a long running nested thread - possibly blocking joining of the parent thread. But that's not the case in python.
It would be a better design not to call processing_executor.submit from inside thread from a request_reading_executor pool, but abstractions around BaseServer don't allow it. 
Ideally we should use an async server like tornado to handle these requests, but that's for another PR :)

alesnovak-s1 replied:
> To me it seems like that this should live on the syslog monitor `stop()` method and not inside the MonitorsManager itself.

Looking more into the code of syslog monitors, I've realized you're right.
Since SyslogMonitor is in charge of starting and stopping all Syslog Servers, it makes sense for it to be responsible for creation and shutting down of the pools.

alesnovak-s1 replied:
In the end it really needs to be in the Monitor Manager, I understood the Syslog Monitor wrong. It starts servers for one monitor configuration only.

alesnovak-s1 commented:
@tomaz-s1 
I have added configuration, changed tests to follow the new api of some of the classes (Syslog Request and Parser).
And now to ensure the request data is processed in the correct order, the worker submitted to the processing executor gets his data from a Queue (one queue and one worker per request). 
Please review again.

tomaz-s1 commented:
@alesnovak-s1 Thanks.

Have we done any end to end testing with agent running multiple monitor instances and producer sending data in parallel (e.g. using https://github.com/scalyr/syslog-collector/tree/main/data-and-load-generator tool)?

tomaz-s1 commented on scalyr_agent/builtin_monitors/syslog_monitor.py:
```diff
@@ -570,77 +583,110 @@ class SyslogUDPHandler(six.moves.socketserver.BaseRequestHandler):
         except socket.error as e:
             if e.errno == errno.EAGAIN:
-                return None
+                raise SocketNotReadyException(e)
```
Just curious why this change is needed? In the past, we had bug related to this code path throwing, but this could have changed with the multi threading model.

alesnovak-s1 replied:
Before the method returned None for both cases - when data on socket were not ready and when the socket was closed.
And you had to guess what happened. 
The exceptions are path of the code flow and are expected and catched right away, avoiding any propagating.

alesnovak-s1 replied:
And I don't like returning None when something unexpected happened. It's quite error-prone.

tomaz-s1 reviewed (approved):
LGTM, but I would feel more comfortable if we performed more end to end testing with multiple monitor instances and highly parallel producer.

alesnovak-s1 commented:
> @alesnovak-s1 Thanks.
> 
> Have we done any end to end testing with agent running multiple monitor instances and producer sending data in parallel (e.g. using https://github.com/scalyr/syslog-collector/tree/main/data-and-load-generator tool)?

Yes, I have a config with multiple Syslog (~15) monitors configured with both udp and tcp ports.
I was using loggen and syslog-loadgen and sending tons of data to all the ports in parallel, several processes per port.
